prov/mrail/src/mrail_init.c

```diff
@@ -157,110 +154,132 @@ int mrail_get_core_info(uint32_t version, const char *node, const char *service,
+		assert(fi->tx_attr->iov_limit);
+		fi->tx_attr->iov_limit--;
+
+		/* Claiming messages larger than FI_OPT_BUFFERED_LIMIT would consume
+		 * a scatter/gather entry for mrail_hdr
+		 */
+		fi->rx_attr->iov_limit--;
+
+		if (fi->tx_attr->inject_size < sizeof(struct mrail_hdr))
+			fi->tx_attr->inject_size = 0;
+		else
+			fi->tx_attr->inject_size -= sizeof(struct mrail_hdr);
+
+		if (hints && hints->tx_attr && (hints->tx_attr->op_flags & FI_COMPLETION))
+			fi->tx_attr->op_flags |= FI_COMPLETION;
+
+		// TODO set src_addr to FI_ADDR_STRC address
+
+		for (i = 0; i < mrail_num_info; i++) {
+			if (!strcmp(core_info->fabric_attr->prov_name,
+			    mrail_info_vec[i]->fabric_attr->prov_name)) {
+				fi = mrail_info_vec[i];
+				while (fi->next)
+					fi = fi->next;
+				fi->next = core_info;
+				fi = fi->next;
+				core_info = core_info->next;
+				fi->next = NULL;
+				break;
+			}
+		}
+		if (!mrail_info_vec[i]) {
+			assert(mrail_num_info < MRAIL_MAX_INFO);
+			mrail_info_vec[i] = core_info;
+			core_info = core_info->next;
+			mrail_info_vec[i]->next = NULL;
+			mrail_num_info++;
+		}
 	}
-	fi->domain_attr->name = strdup(mrail_info.domain_attr->name);
-	if (!fi->domain_attr->name) {
+
+	*info = fi_dupinfo(mrail_info_vec[0]);
+	if (!*info) {
 		ret = -FI_ENOMEM;
-		goto err2;
+		goto err;
 	}
-	fi->ep_attr->protocol 		= mrail_info.ep_attr->protocol;
-	fi->ep_attr->protocol_version 	= mrail_info.ep_attr->protocol_version;
-	fi->fabric_attr->prov_version	= FI_VERSION(MRAIL_MAJOR_VERSION,
-						     MRAIL_MINOR_VERSION);
-	fi->domain_attr->mr_key_size 	= mr_key_size;
-	fi->domain_attr->mr_mode        |= FI_MR_RAW;
-
-	/* Account for one iovec buffer used for mrail header */
-	assert(fi->tx_attr->iov_limit);
-	fi->tx_attr->iov_limit--;
 
-	/* Claiming messages larger than FI_OPT_BUFFERED_LIMIT would consume
-	 * a scatter/gather entry for mrail_hdr */
-	fi->rx_attr->iov_limit--;
-
-	if (fi->tx_attr->inject_size < sizeof(struct mrail_hdr))
-		fi->tx_attr->inject_size = 0;
-	else
-		fi->tx_attr->inject_size -= sizeof(struct mrail_hdr);
-
-	if (hints && hints->tx_attr && (hints->tx_attr->op_flags & FI_COMPLETION))
-		fi->tx_attr->op_flags |= FI_COMPLETION;
-
-	// TODO set src_addr to FI_ADDR_STRC address
-	fi->next = *info;
-	*info = fi;
+	for (i = 0; i < mrail_num_info; i++) {
+		if (i == 0) {
+			tail = fi_dupinfo(mrail_info_vec[i]);
+			*info = tail;
+		} else {
+			tail->next = fi_dupinfo(mrail_info_vec[i]);
+			tail = tail->next;
+		}
+		if (!tail) {
+			ret = -FI_ENOMEM;
+			goto err;
+		}
+		tail->next = NULL;
 
-	mrail_info_vec[mrail_num_info] = mrail_dupinfo(*info);
-	if (!mrail_info_vec[mrail_num_info])
-		goto err2;
+		fi = mrail_info_vec[i];
 
-	mrail_num_info++;
+		while (fi) {
+			snprintf(fi->fabric_attr->prov_name+
+				 strlen(fi->fabric_attr->prov_name),
+				 FI_NAME_MAX,
+				 ";%s",
+				 mrail_prov.name);
+			fi = fi->next;
+		}
+	}
 
 	return 0;
-err2:
-	fi_freeinfo(fi);
-err1:
+
+err:
+	fi_freeinfo(core_info);
```

Comment:
Please try to keep functions to about 1 screen length in size.  Create sub-routines where needed to help with readability.